### PR TITLE
Kmer mapping diagnostic

### PR DESCRIPTION
**Pull request by zrm22**
## Description

Updates fixing the BuildKmerIndex ULong issue and making it only keep track of Kmers which hit less than .5 * numSampleGametes to only keep informative Kmers in the map.

It will still need some RAM efficiency improvements to run on smaller machines.

## Type of change

_What type of changes does your code introduce? Put an `x` in boxes that apply._

- [x] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Checklist:

- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have updated relevant documentation

## Changelog entry

_Please add a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
Updating Build Kmer index to keep track of more Kmers to improve mapping rates.
```

### CODE REVIEW

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/BuildKmerIndex.kt [hunk 1]
```diff
@@ -123,23 +138,22 @@ class BuildKmerIndex: CliktCommand(help="Create a kmer index for a HaplotypeGrap
...
+                          hashMask: Long = 3, hashFilterValue:Long = 1) : Pair<Long2ObjectOpenHashMap<Set<String>>,LongOpenHashBigSet> {
         //keepMap is a map of hash -> Set of haplotype ids
-        val keepMap = Long2ObjectOpenHashMap<Set<String>>()
+        //TODO Change this to be more RAM efficient.  Probably need something like this:
+        // Map<Long, Set<Pair<refRangeIdx, hapIdBitset>>
+        val keepMap = Long2ObjectOpenHashMap<MutableSet<String>>(500_000_000)
 
         //discardSet is a Set of hashes
-        val discardSet = LongOpenHashSet()
+        val discardSet = LongOpenHashBigSet(500_000_000)
```
Comment (lynnjo) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/BuildKmerIndex.kt:149]: Were you going to change this from LongOpenHashBigSet to LongOpenHashSet ?

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/BuildKmerIndex.kt [hunk 2]
```diff
@@ -123,23 +138,22 @@ class BuildKmerIndex: CliktCommand(help="Create a kmer index for a HaplotypeGrap
...
         val sampleGametes = graph.sampleGametesInGraph()
 
         val contigRangesMap = graph.rangesByContig()
 
-        //needed for diagnostics
-        val refRangeToIndexMap = graph.refRangeToIndexMap()
-        val hapidToRefrangeMap = graph.hapIdToRefRangeMap()
-
-        val maxHapsToKeep = sampleGametes.size * 2
+        //This is a parameter we are currently testing
```
Comment (lynnjo) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/BuildKmerIndex.kt:155]: Maybe a slightly expanded comment as to why we only want to keep sampleGametee.size * .5 vs the previous *2.  A general comment on why a smaller proportion of them is better than a greater proportion

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/MapKmers.kt [hunk 3]
```diff
@@ -96,7 +96,8 @@ class MapKmers : CliktCommand(help="Map Kmers to the pangenome reference") {
 
     val minProportionSameReferenceRange by option(help = "Minimum proportion of the read that must align to the same reference range.")
         .double()
-        .default(0.9)
+//        .default(0.9)
+        .default(0.0)
```
Comment (lynnjo) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/MapKmers.kt:101]: Add a comment on why user would want this value to be low vs high, 0 vs .9

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/QcReadMapping.kt [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+import com.github.ajalt.clikt.parameters.options.option
+import com.github.ajalt.clikt.parameters.options.required
+import com.github.ajalt.clikt.parameters.types.int
+import htsjdk.samtools.fastq.FastqReader
+import net.maizegenetics.phgv2.api.HaplotypeGraph
+import net.maizegenetics.phgv2.pathing.AlignmentUtils.Companion.convertRefRangeToIdBitsetMap
+import net.maizegenetics.phgv2.pathing.AlignmentUtils.Companion.rangeHapidMapFromKmerHash
+import java.io.File
+import kotlin.math.min
+
```
Comment (lynnjo) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/QcReadMapping.kt:16]: Add a header comment summarizing what this class does.

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/BuildKmerIndex.kt [hunk 5]
```diff
@@ -367,28 +317,26 @@ class BuildKmerIndex: CliktCommand(help="Create a kmer index for a HaplotypeGrap
                     //start using kmers starting with the 32nd nucleotide
                     for (nucleotide in sequence.subSequence(31 until sequence.length)) {
                         previousHash = updateKmerHashAndReverseCompliment(previousHash, nucleotide)
-                        val minHash = min(previousHash.first, previousHash.second)
+                        val minHash = min(previousHash.first.toULong(), previousHash.second.toULong()).toLong()
```
Comment (smm477) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/BuildKmerIndex.kt:320]: I would explain why you go from unsigned L back to signed L.

**Review by smm477 — APPROVED** [review]

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/QcReadMapping.kt [hunk 6]
```diff
@@ -0,0 +1,114 @@
+package net.maizegenetics.phgv2.pathing
```
Comment (tcasstevens) [inline src/main/kotlin/net/maizegenetics/phgv2/pathing/QcReadMapping.kt:1]: Shouldn't this be in phgv2.cli?

**Review by tcasstevens — APPROVED** [review]

File: src/main/kotlin/net/maizegenetics/phgv2/pathing/QcReadMapping.kt (hunk 6 above)
Comment (zrm22) [reply src/main/kotlin/net/maizegenetics/phgv2/pathing/QcReadMapping.kt:1]: No, all the other path finding Clikt commands are in their own package so not everything is in the cli package.  This is related to pathing so I put it in there.

**Review by aberthel — APPROVED** [review]

**Pull request merged**